### DISCUSSION TITLE
bug: problem handling Perplexity Online models via OpenRouter

### DISCUSSION THREAD

**Discussion by neuromechanist** [open-webui/open-webui#7043] [Issues]
# Bug Report

## Installation Method
I have installed OpenWebUI using Docker.

## Environment
- **Open WebUI Version:** v0.3.35
- **Ollama (if applicable):** [none]
- **Operating System:** Remote Mac
- **Browser (if applicable):** Safari 18.02, Edge 131

**Confirmation:**
- [x] I have read and followed all the instructions provided in the README.md.
- [x] I am on the latest version of both Open WebUI and Ollama.
- [ ] I have included the browser console logs.
- [ ] I have included the Docker container logs.
- [x] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior:
I expect to receive a response when making API calls through OpenWebUI for the Perplexity Online models (Sonar 8B, 70B, 405B), similar to how other models work (e.g., OpenAI, Anthropic).

## Actual Behavior:
For the Perplexity Online models, I am not receiving any response when making API calls through OpenWebUI. This issue started occurring two days ago, and updating OpenWebUI did not resolve the problem. However, the Chat interface on the OpenRouter page is responding correctly. Handling other models through OpenRouter and OpenAI are ok.

## Description
**Bug Summary:**
OpenWebUI is not returning responses for Perplexity Online models, despite other models functioning correctly.

## Reproduction Details
**Steps to Reproduce:**
1. Install OpenWebUI using Docker on a remote Mac.
2. Attempt to make an API call to the Perplexity Online models through OpenWebUI via OpenRouter.
3. Observe that no response is returned.
4. Test the same API call through the Chat interface on the OpenRouter page, which returns a response as expected.

## Logs and Screenshots
**Browser Console Logs:**
[Link to a Chat JSON](https://drive.google.com/file/d/12v670LVMnrvLhUYbOoILrss50cSjzf72/view?usp=sharing) file

## Additional Information
The issue seems to be linked to a recent change, as mentioned by the OpenRouter team, involving the addition of a citations field in the response. It appears that OpenWebUI may need to handle this new field appropriately. This issue was not present before two days ago and has affected both remote machines and the server.

## Note
Please let me know if I can provide more info. Thank you for your help!

**Comment by neuromechanist** [comment]
I ended up making a new pipline, since the current pipeline example does not work either. Here you can find the pipeline:
https://gist.github.com/neuromechanist/0fdc06d96840d2f51dacd195c2edefa0


The good thing is that references are now included in the response:


> user > Hello
> 
> Sonar Large >
> The term "Hello" can refer to several different things, depending on the context:
> 
> 1. **Lionel Richie's Song**: "Hello" is a famous song by Lionel Richie, released in 1984. The song is known for its catchy melody and heartfelt lyrics about reconnecting with an old flame. The official music video for the song has been remastered in HD and is available on YouTube[1].
> 
> 2. **Adele's Song**: Adele also has a song titled "Hello," which was released in 2015 as part of her album "25." This song is a soulful ballad about regret and longing, and it became a huge commercial success[2].
> 
> 3. **HelloFresh**: HelloFresh is a meal kit delivery service that provides fresh, delicious meals with easy recipes. It offers a variety of meal plans and allows customers to skip weeks, pause, or cancel their subscriptions at any time[3].
> 
> 4. **Hello Alice**: Hello Alice is a platform designed to help small businesses grow by providing personalized growth plans and financial health scores. It aims to position businesses for success and offers rewards for growth[4].
> 
> 5. **Super Simple Songs**: "Hello!" is also the title of a fun and energetic song from Super Simple Songs, designed for children to learn about greetings and social interactions[5].
> 
> Each of these "Hellos" serves a different purpose, whether it's music, food delivery, business growth, or children's education.
> 
> References:
> [1] https://www.youtube.com/watch?v=mHONNcZbwDY
> [2] https://www.youtube.com/watch?v=yjo_aXygRDI
> [3] https://www.hellofresh.com
> [4] https://helloalice.com
> [5] https://supersimple.com/song/hello/

**Comment by CSAnetGmbH** [comment]
Thank you very much, but in my Openwebui when i Click Save in Pipelines an Error Pops UP: No Valves....
What can this cause ?

**Comment by CSAnetGmbH** [comment]
@neuromechanist Hello, can you help me ? in Openwebui, when i click Save in Pipelines, an Error Pops UP: No Valves....
What can this cause ? How can i solve the Problem. 

Thank you

**Comment by neuromechanist** [comment]
Hi,
I am not very familiar with this error, but it seems that the Python file could not be detected. Probably try to use the path to file from my PR ([here is the link](https://github.com/neuromechanist/pipelines/blob/patch-2/examples/pipelines/providers/perplexity_manifold_pipeline.py)), so you can directly pass the URL to OpenWebUI:
<img width="1434" alt="image" src="https://github.com/user-attachments/assets/72b36296-05f3-4890-bf5f-6588fd4b962d">

Then click on the download icon, you should see on the bottom a text field appearing asking you to paste your Perplexity API key.

I used two different versions of OpenWebUI (0.3.21, and 0.3.35) and also used the upload local Python copy and the Github link, all worked as expected.

**Reply by CSAnetGmbH** [reply]
Thank you so much for Your Reply. I have tried this but i don´t see the Bottom Fields "PIPELINE VALVES" and "Perplexity Api Key:
![grafik](https://github.com/user-attachments/assets/a1924869-45b3-4168-812c-0c97db4e8322)

The interesting point is: When I try to install the Pipeline from 4 Month ago : https://github.com/neuromechanist/pipelines/blob/d71dab0906918238664f0cb0cf9476fda07545b2/examples/pipelines/providers/perplexity_manifold_pipeline.py
This old Version works.


I use Openwebui v0.3.35

Can you imagine, what this Error cause ? It must be something different in the 2 Versions ?!

Thanks a lot !

Ralf

**Reply by CSAnetGmbH** [reply]
Any Idea ?

**Reply by neuromechanist** [reply]
The screenshot you showed has both fields. Is the image you included a part of the screenshot that I attached before?
If you can't see these fields, it might mean that the pipeline is not added, or that pipeline is not running in the background.

Alternatively, you can put the Pythoin file directly inside the Pipeline directory. Mine currently shows this for example, as I have both Antroplic and Perplexity pipelines:
<img width="271" alt="image" src="https://github.com/user-attachments/assets/3360b0a8-e848-4f30-8b94-b1a1aea70b03">


After you copy the Python file here, you should be able to find that pipeline in the dropdown menu after restarting the Pipeline engine.

**Reply by CSAnetGmbH** [reply]
Hello Thank you, I have made this Screenshot of my own, its the old Version of the Perplexity Pipeline.

The new Version will not work .-(
What i have tried:
Delete the old Version
Tryed to install it via Pipline Upload , Response in Red ( No Valves to update ) / won´t Work
Tryed to install it via Github Url, Response in Red ( No Valves to update ) / won´t Work

Tried to put the File in the Directory of Pipelines , it isnt shown in the Dropdown Menu

Then I tried the old one [reddiedev](https://github.com/neuromechanist/pipelines/commits?author=reddiedev)
reddiedev , it works fine ( but not the Online Model )

What i see in your Screenshot, there is another Folder called Perplexity Pipeline, I don´t have this. Could this be the Reason ?

Thanks a lot for your Help
Ralf

**Reply by CSAnetGmbH** [reply]
Hello, I have found the Problem. It is the command "pass" wich is not implemented in your Code :-) Now it works fine !

Regards
Ralf

**Reply by neuromechanist** [reply]
Thanks for the effort. I never would have thought that the pass argument would be needed, it is supposed to be a placeholder.
I added the `pass` statements for consistency.

**Comment by neuromechanist** [comment]
Perplexity Sonar Online models are being discontinued in two weeks. @harry71019 updated the pipeline with the new model handles. You can download the updated pipeline from [this gist](https://gist.github.com/harry71019/d5672b03b4e556d305ab38671b9b15f6).

Also, the [pricing](https://docs.perplexity.ai/guides/pricing) is changing as well, it seems that they are getting rid of Sonar Small, and the cheapest is the Sonar model now, which has the same cost as the legacy Sonar Large.

**Comment by cpbotha** [comment]
Just in case anyone else would like to hack this right into Open WebUI itself, i.e. without using a pipeline, I just got this working by making three small changes to `backend/open_webui/utils/middleware.py`.

Read here for a bit more detail:
https://charlbotha.com/til/Hack-Open-WebUI-to-display-Perplexity-Sonar-citations-list

... or apply this diff and restart:

```diff
diff --git a/backend/open_webui/utils/middleware.py b/backend/open_webui/utils/middleware.py
index 4070bc697..22914bffc 100644
--- a/backend/open_webui/utils/middleware.py
+++ b/backend/open_webui/utils/middleware.py
@@ -1577,6 +1577,8 @@ async def process_chat_response(
 
                     response_tool_calls = []
 
+                    # cpbotha: init citations list
+                    citations_list = None
                     async for line in response.body_iterator:
                         line = line.decode("utf-8") if isinstance(line, bytes) else line
                         data = line
@@ -1617,6 +1619,12 @@ async def process_chat_response(
                                         },
                                     )
                                 else:
+                                    # cpbotha: check here if there are any citations and store them for rendering
+                                    if data.get("citations", None):
+                                        citations_list = "\n\n**Citations**\n" + "\n".join(
+                                            [f"{i + 1}. {c}" for i, c in enumerate(data.get("citations", []))]
+                                        )
+
                                     choices = data.get("choices", [])
                                     if not choices:
                                         error = data.get("error", {})
@@ -1827,6 +1835,20 @@ async def process_chat_response(
                                 log.debug("Error: ", e)
                                 continue
 
+                    # cpbotha: if we have a citations list, add it to the content[_blocks]
+                    if citations_list is not None:
+                        # follow exactly the same append to content AND content_blocks pattern as above for delta.get("content")
+                        content = f"{content}{citations_list}"
+                        if not content_blocks:
+                            content_blocks.append(
+                                {
+                                    "type": "text",
+                                    "content": "",
+                                }
+                            )
+
+                        content_blocks[-1]["content"] += f"{citations_list}"
+
                     if content_blocks:
                         # Clean up the last text block
                         if content_blocks[-1]["type"] == "text":
```
